http2: minor cleanup of compat, tests

apapirovski opened the pull request:
This is a bit of a hodgepodge of stuff that's too small for its own PRs. Here's the full list:

- Remove path getter and setter, related to https://github.com/expressjs/express/pull/3390.

- ~~Add validation for header names and values, as per the spec. @mcollina I would like your input on this as doing this in compat means that users can still set invalid headers through the API in core. It doesn't feel like we want this in two places but if we only validate in `mapToHeaders` then methods like `setHeader` won't throw when they probably should. I'm also not sure if I should be importing code from http or making a copy in http2 utils — not sure how node would usually handle this.~~

- A few of the validations within if statements were moved up, such as checking for `stream.destroyed` in `kBeginSend` to avoid executing unnecessary code.

- Remove unused properties in kState.

- Remove all the extra code for creating kHeaders & kTrailers and instead just create them in the constructor. I ran a benchmark and this had no performance impact but it keeps the code a lot cleaner since we don't have to keep constantly checking for it.

- Cleanup unnecessary variable assignments throughout to match the majority of the code in h2 and node in general.

Let me know if there's anything I can adjust! Thanks.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
http2, test

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -155,23 +167,17 @@ class Http2ServerRequest extends Readable {
   get closed() {
-    const state = this[kState];
-    return Boolean(state.closed);
+    return this[kState].closed;
```
I think `kState` can be nulled, shouldn't we protect this?

apapirovski replied:
I don't think there's ever a situation where that can happen (I reviewed all of the code in compat) and since it's accessed via a Symbol, users would have to really go out of their way to null it.

mcollina replied:
Ok, good catch!

mcollina commented on lib/internal/http2/compat.js:
```diff
@@ -201,7 +207,7 @@ class Http2ServerRequest extends Readable {
 
   get socket() {
-    return this.stream.session.socket;
+    return this[kStream].session.socket;
```
`this[kStream]` can be nulled, shouldn't we protect this?

apapirovski replied:
Yeah, I agree. It wasn't protected before so I wasn't sure if there was a reason for it or not.

On that note, hapi uses this to set their internal processing flag and since our stream reference isn't accessible after 'finish', it throws. Should we maybe store a reference to the session directly and not null it on finish? It could lead to some garbage collection issues but on the flipside it's the only issue I found with hapi compatibility.

mcollina replied:
@apapirovski I don't think what HAPI is actually doing in HTTP1 could be ported 1-1 to HTTP2. I've added this to be compatibile but for all intent and purposes this should not be used: the socket is controlled by the C++/C layers. 

@jasnell what do you think?

apapirovski replied:
Yeah, it's a bit weird because they don't actually seem to use the socket based on my quick review of the code — maybe they did in the past. Now they just store a flag on it. When I built a version that retained the socket reference then my test site worked fine with h2 on hapi. But I agree with you re: not using it.

I guess it's all about how much we're willing to compromise for compatibility.

apapirovski replied:
Just a thought but if we do something like `process.nextTick(() => (this[kStream] = undefined));` then we get the best of both worlds. It passes all our tests, makes hapi work and gets rid of the stream reference after the user finishes processing the 'finish' event. That's probably the most h1 compatible solution.

mcollina commented on test/parallel/test-http2-compat-serverrequest-headers.js:
```diff
@@ -23,7 +23,6 @@ server.listen(0, common.mustCall(function() {
 
     assert.strictEqual(request.method, expected[':method']);
     assert.strictEqual(request.scheme, expected[':scheme']);
-    assert.strictEqual(request.path, expected[':path']);
```
I think we should add an assertion that `request.path`  is `undefined` for express compatibility.

apapirovski replied:
Good thinking, will update.

apapirovski commented on lib/internal/http2/compat.js:
```diff
@@ -26,21 +45,21 @@ let statusMessageWarned = false;
+    case HTTP2_HEADER_METHOD:
+    case HTTP2_HEADER_PATH:
+    case HTTP2_HEADER_AUTHORITY:
+    case HTTP2_HEADER_SCHEME:
```
Yeah, I just wanted to have this similar to core where everything is a constant. I'll add the comments for sure.

apapirovski commented:
Ok, new committed added. Mostly just addressed feedback and added a couple of extra protections that were missing. The one debatable change is, as I mentioned above, setting kStream to undefined on nextTick instead of immediately. Open to feedback on that.

apapirovski commented:
Moved the header validation out for now. Can revisit in a separate PR.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10041/

mcollina commented:
This does not land cleanly on master, would you mind rebasing and squashing?

apapirovski commented:
Rebased and squashed, also updated the commit message to be more descriptive + added links. Should be good to go.

mcollina commented:
Landed as c981483686be11980908ebebe77723ae95d05b86
